src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
     }
 }
 
-int ompl::geometric::RRTstar::pruneTree(const base::Cost pruneTreeCost)
+int ompl::geometric::RRTstar::pruneTree(const base::Cost& pruneTreeCost)
 {
-    const int tree_size = nn_->size();
-    pruneScratchSpace_.newTree.reserve(tree_size);
-    pruneScratchSpace_.newTree.clear();
-    pruneScratchSpace_.toBePruned.reserve(tree_size);
-    pruneScratchSpace_.toBePruned.clear();
-    pruneScratchSpace_.candidates.clear();
-    pruneScratchSpace_.candidates.push_back(startMotion_);
-    std::size_t j = 0;
-    while (j != pruneScratchSpace_.candidates.size())
+    // Variable
+    // The percent improvement (expressed as a [0,1] fraction) in cost
+    double fracBetter;
+    // The number pruned
+    int numPruned = 0;
+
+    if (opt_->isFinite(prunedCost_))
     {
-        Motion *candidate = pruneScratchSpace_.candidates[j++];
-        if (opt_->isCostBetterThan(pruneTreeCost, costToGo(candidate)))
-            pruneScratchSpace_.toBePruned.push_back(candidate);
-        else
+        fracBetter = std::abs((pruneTreeCost.value() - prunedCost_.value())/prunedCost_.value());
+    }
+    else
+    {
+        fracBetter = 1.0;
+    }
+
+    if (fracBetter > pruneThreshold_)
+    {
+        // We are only pruning motions if they, AND all descendents, meet the pruning criterion
+        // The easiest way to do this is to find all the leaves of the tree that meet the pruning criterion, and ascend up the tree from each leaf until a motion is found that should not be pruned
+
+        // Variable
+        // The list of motions in the NN structure, initially full of nulls
+        std::vector<Motion*> treeVertices(nn_->size(), NULL);
+        // The list of leaves to prune
+        std::list<Motion*> leavesToPrune;
+
+        // Get the list from the NN structure
+        nn_->list(treeVertices);
```

Comment:
I didn't want to start from startMotion_, since if we had multiple start states, we could have grown multiple independent trees.